File: DP/322. Coin Change/main.py
#peek
#buttom up
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        max_amount = amount + 1
        amount_lst = [max_amount] * max_amount
        amount_lst[0] = 0 # no coin in list

        for i in range(1, max_amount):
            for c in coins:
                if c <= i:
                    amount_lst[i] = min(amount_lst[i], amount_lst[i-c] + 1)

        return amount_lst[amount] if amount_lst[amount] < max_amount else -1

#top down
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        max_amount = amount + 1
        amount_lst = [0] * max_amount

        def search_amount_lst(amount, lst):
            if amount < 0: return -1
            if amount == 0: return 0
            if lst[amount] != 0: return lst[amount]
            minumun = max_amount
            for c in coins:
                ret = search_amount_lst(amount-c, lst)
                if ret >= 0 and ret < minumun:
                    minumun = ret + 1
            
            lst[amount] = -1 if minumun == max_amount else minumun
            return lst[amount]
                
        return search_amount_lst(amount, amount_lst)

File: DP/322. Coin Change/go/main.go
package main

import "math"

// Top Down - recursion
// m = amount, n = len(coins)
// time complexity: O(n*m)
// space complexity: O(m)
func coinChangeRecursive(coins []int, amount int) int {
	if amount == 0 {
		return 0
	}
	memo := make([]int, amount)

	dfs(&memo, coins, amount)

	return memo[amount-1]
}

func dfs(memo *[]int, coins []int, amount int) int {
	if amount == 0 {
		return 0
	}
	if amount < 0 {
		return -1
	}
	if (*memo)[amount-1] != 0 {
		return (*memo)[amount-1]
	}

	min := math.MaxInt
	for _, c := range coins {
		ret := dfs(memo, coins, amount-c)
		if ret != -1 && min > ret {
			min = ret
		}
	}

	if min == math.MaxInt {
		(*memo)[amount-1] = -1
	} else {
		(*memo)[amount-1] = min + 1
	}
	return (*memo)[amount-1]
}

// buttom up - tabulation
// m = amount, n = len(coins)
// time complexity: O(m*n)
// space complexity: O(m)
func coinChange(coins []int, amount int) int {
	if amount == 0 {
		return 0
	}
	table := make([]int, amount+1)
	for i := range table {
		table[i] = -1
	}
	table[0] = 0

	for i := 0; i < len(table); i++ {
		if table[i] == -1 {
			continue
		}
		for _, c := range coins {
			cur := i + c
			if cur < len(table) {
				if table[cur] == -1 || table[cur] > table[i]+1 {
					table[cur] = table[i] + 1
				}
			}
		}
	}

	if table[len(table)-1] == 0 {
		return -1
	}
	return table[len(table)-1]
}
